hudi-common/src/main/java/org/apache/hudi/common/serialization/RecordSerializer.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.common.serialization;
+
+import java.io.Serializable;
+
+/**
+ * Serializer for engine-specific record, e.g., {@code InternalRow} in Spark,
+ * {@code RowData} in Flink and {@code IndexedRecord} for Avro.
+ */
+public interface RecordSerializer<T> extends Serializable {
```

Comment:
This is very similar to the interface we already have, `CustomSerializer`. I'm guessing we can't consolidate due to the deserializer needing a schema? The records in the spillable map should all have the same schema since we read the log records with the projected schema from the log files so maybe we don't need that `schemaId`?

Reply:
> The records in the spillable map should all have the same schema

That's what I originally thought, but unfortunately not for MDT scenario. Currently, merging result of two MDT log records  is always `HoodieMetadataRecord`, which has a definite schema `HoodieMetadataRecord#SCHEMA$`, which can be different with the required schema of FG reader. So we decode schema using `schemaId` in `BufferedRecord` to make sure the schema is matching the Avro record itself.

Reply:
I ran into this with the metadata compaction implementation as well. There can be runtime issues if we have mixed schemas with our records though so we need some way to standardize how to handle this.